modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java

```diff
@@ -313,9 +323,55 @@ public boolean getShouldStoreResult() {
         return shouldStoreResult;
     }
 
+    /**
+     * The number of workers this task should be divided into. Defaults to 1 meaning the task is single threaded.
+     */
+    public Self setWorkers(int workers) {
+        if (workers < 1) {
+            throw new IllegalArgumentException("[workers] must be at least 1");
+        }
+        this.workers = workers;
+        return self();
+    }
+
+    /**
+     * The number of workers this task should be divided into. Defaults to 1 meaning the task is single threaded.
+     */
+    public int getWorkers() {
+        return workers;
+    }
+
+    /**
+     * Build a new request for a slice of the parent request.
+     */
+    abstract Self forSlice(TaskId slicingTask, SearchRequest slice);
+
+    /**
+     * Setup a clone of this request with the information needed to process a slice of it.
+     */
+    protected Self doForSlice(Self request, TaskId slicingTask) {
+        request.setAbortOnVersionConflict(abortOnVersionConflict).setRefresh(refresh).setTimeout(timeout)
+                .setWaitForActiveShards(activeShardCount).setRetryBackoffInitialTime(retryBackoffInitialTime).setMaxRetries(maxRetries)
+                // Parent task will store result
+                .setShouldStoreResult(false)
+                // Split requests per second between all slices
+                .setRequestsPerSecond(requestsPerSecond / workers)
+                // Size is split between workers. This means the size might round down!
+                .setSize(size == SIZE_ALL_MATCHES ? SIZE_ALL_MATCHES : size / workers)
+                // Sub requests don't have workers
+                .setWorkers(1);
+        // Set the parent task so this task is cancelled if we cancel the parent
+        request.setParentTask(slicingTask);
+        // TODO We could probably yank refresh out of this and move it to the parent request....
+        return request;
+    }
+
     @Override
     public Task createTask(long id, String type, String action, TaskId parentTaskId) {
-        return new BulkByScrollTask(id, type, action, getDescription(), parentTaskId, requestsPerSecond);
+        if (workers > 1) {
```

Comment:
A little bit of task fun here. Since each request is its own individual task that can be addressed via the task API we use this `ParentBulkByScrollTask` to merge all of the tasks together for viewing, rethrottling, cancellation, and building the result. One issue with this approach is that the children tasks have to be on the same node as the parent task. That limits some optimization we could make in the future but I figure if we want to make those optimizations we can rework this some.